go/vt/vtgate/executor_scatter_stats.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2019 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vtgate
+
+import (
+	"fmt"
+	"html/template"
+	"net/http"
+	"time"
+
+	"vitess.io/vitess/go/vt/logz"
+
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vtgate/engine"
+)
+
+type statsResults struct {
+	TotalReadOnlyTime  time.Duration
+	TotalScatterTime   time.Duration
+	PercentTimeScatter float64
+	Items              []*statsResultItem
+}
+
+// using this intermediate data structure to make testing easier
+type statsResultItem struct {
+	Query                  string
+	AvgTimePerQuery        time.Duration
+	PercentTimeOfReads     float64
+	PercentTimeOfScatters  float64
+	PercentCountOfReads    float64
+	PercentCountOfScatters float64
+	From                   string
+	Count                  uint64
+}
+
+func (e *Executor) gatherScatterStats() (statsResults, error) {
+	scatterExecTime := time.Duration(0)
+	readOnlyTime := time.Duration(0)
+	scatterCount := uint64(0)
+	readOnlyCount := uint64(0)
+	totalExecTime := time.Duration(0)
+	totalCount := uint64(0)
+
+	plans := make([]*engine.Plan, 0)
+	routes := make([]*engine.Route, 0)
+	// First we go over all plans and collect statistics and all query plans for scatter queries
```

Comment:
FYI, at some point in the future I may look into generalizing this to take query parameters for plan type, vindex name, and/or vindex type, or potentially even a substring string search on the query.

It's possible to do all that today from the command line with `curl` + `jq`, but if we have a slightly smarter plan viewer/browser that'd make life easier